### DISCUSSION TITLE
Support exponential notation for integer literals (e.g. 1e9 for 1 billion)

### DISCUSSION THREAD

**Discussion by GSPP** [dotnet/csharplang#5641] [Language Ideas]
"e" literals are always interpreted as floating-point. It would be convenient if they could be used to specify integers.

    int x = 1e9;

C# has gotten a lot of target-typing lately. Maybe this strategy can be used to specify this cleanly.

**Comment by DavidArno** [comment]
Would this feature really be useful in all but a tiny number of cases for a small number of developers? Afterall, I might want to specify the mass of the Titanic in Kg:
```c#
int massTitanic1 = 4.6e7; // how will the compiler know it's OK to convert that to an int?
int massTitanic2 = 46e6; // And is this worth implementing?
var massTitanic3 = (int)4.6e7; // When I can already do this and I can use var here as a bonus.
var massTitanic4 = 46_000_000; // And really, are any of the above better than this? 
```

**Reply by huoyaoyuan** [reply]
Given the range of integers, using literal delimiter isn't too verbose.

**Reply by Xyncgas** [reply]
> Would this feature really be useful in all but a tiny number of cases for a small number of developers? Afterall, I might want to specify the mass of the Titanic in Kg:
> 
> ```cs
> int massTitanic1 = 4.6e7; // how will the compiler know it's OK to convert that to an int?
> int massTitanic2 = 46e6; // And is this worth implementing?
> var massTitanic3 = (int)4.6e7; // When I can already do this and I can use var here as a bonus.
> var massTitanic4 = 46_000_000; // And really, are any of the above better than this? 
> ```

coming in hot

**Reply by CyrusNajmabadi** [reply]
@Xyncgas Please try to keep a good signal to noise ratio in your posts.  Thanks! :)

**Reply by GSPP** [reply]
These are not terribly convincing, I admit. I was thinking essentially of 1e3, 1e6 and 1e9... I just happened to need 1e9, and switch to 1e6 (because I had confused seconds and milliseconds). That scenario would be a lot more convenient with "e syntax".

**Reply by RikkiGibson** [reply]
> ```cs
> int massTitanic1 = 4.6e7; // how will the compiler know it's OK to convert that to an int?
> ```

We do check ranges when a constant-valued expression is converted to int, for example. It doesn't seem *impossible* to check a constant-valued float or double expression and decide whether it is "lossy" when converting to an integer type. [SharpLab](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYYCWAdgC7oCyAFAJT6HGtQDs6dFlD3VAOhoBDAB4A1YQBsArgFN0YdHAAyDANwsiAX2TagA)
